File: leetcode/easy/1143_longestCommonSubsequence.js
/**
 * @param {string} text1
 * @param {string} text2
 * @return {number}
 */
var longestCommonSubsequence = function(text1, text2) {
    const dp = new Array(text1.length + 1).fill(0).map(() => new Array(text2.length + 1).fill(0))

    for (let i = 1; i <= text1.length; ++i) {
        for (let j = 1; j <= text2.length; ++j) {
            let val = Math.max(dp[i-1][j], dp[i][j-1]);
            if (text1[i - 1] === text2[j - 1]) {
                val = Math.max(dp[i-1][j-1] + 1, val);
            }
            dp[i][j] = val;
        }
    }



    return dp[text1.length][text2.length];
};

/*

  a b c d e
a
c
e



  a b c d e
a 1 1 1 1 1
c 1 1 2 2 2
e 1 1 2 2 3

  a b c d e
c 0 0 1 1 1
c 0 0 1 1 1
e 0 0 1 1 2
e 0 0 1 1 2


*/


const test = () => {
    const params = [
        {
            input: ["abcde", "ace" ],
            output: 3,
        },
        {
            input: ["abc", "abc"],
            output: 3,
        },
        {
            input: ["abc", "def"],
            output: 0,
        },
    ];

    for (const { input, output } of params) {
        const result = longestCommonSubsequence(...input);
        const message = `
            INPUT: ${JSON.stringify(input)}
            OUTPUT: ${output}
            RESULT: ${result}
            `;

        if (JSON.stringify(result) === JSON.stringify(output)) {
            console.log(
                'SUCCESS: \n', message,
            );
        } else {
            console.error('ERROR: \n', message);
        }
    }
};

test();

File: leetcode/medium/1143_longestCommonSubsequence.py
from typing import List
import json
from collections import deque, defaultdict

class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        n = len(text1)
        m = len(text2)
        memo = [[0]*m for i in range(n)]

        for i in range(n):
            for j in range(m):
                maxTop = 0 if i == 0 else memo[i-1][j]
                maxLeft = 0 if j == 0 else memo[i][j-1]
                maxPrev = 0 if (i == 0 or j == 0) else memo[i-1][j-1]

                if text1[i] == text2[j]:
                    maxPrev += 1

                memo[i][j] = max(maxTop, maxLeft, maxPrev)

        return memo[-1][-1]


'''
  a b c d e
a 1 1 1 1 1
c 1 1 2 2 2
e 1 1 2 2 3

'''


def test ():
    params = [
        {
            'input': ["abcde", "ace"],
            'output': 3,
        },
        {
            'input': ["abc", "abc"],
            'output': 3,
        },
        {
            'input': ["abc", "def"],
            'output': 0,
        },
    ]
    solution = Solution()

    for param in params:
        str1, str2 = param['input']
        result = solution.longestCommonSubsequence(str1, str2)
        correct = json.dumps(result) == json.dumps(param['output'])

        msg = 'SUCCESS' if correct else 'ERROR'
        msg += '\n'
        if not correct:
            msg += 'input ' + json.dumps(param['input']) + '\n'
            msg += 'output ' + json.dumps(param['output']) + '\n'
            msg += 'result ' + json.dumps(result) + '\n'

        print(msg)


if __name__ == '__main__':
    test()
